it/src/main/java/com/google/cloud/teleport/it/DataGenerator.java

```diff
@@ -0,0 +1,152 @@
+    return new DataGenerator.Builder(jobName);
+  }
+
+  /**
+   * Executes the data generator using the config provided. If messageLimit is provided, we wait
+   * until the data generator finishes or we reach timeout. If a messageLimit is not provided we
+   * wait until timeout and cancel the data generator.
+   *
+   * <p>Note: This is a blocking call. For backlog tests, start the Data generator before executing
+   * the template under test. For testing a pipeline against live incoming data, execute the data
+   * generator after starting the pipeline under test.
+   *
+   * @param timeout time to wait before cancelling the data generator.
+   * @throws IOException if any errors are encountered.
+   */
+  public void execute(Duration timeout) throws IOException {
+    JobInfo dataGeneratorJobInfo = dataflowClient.launch(PROJECT, REGION, dataGeneratorOptions);
+    assertThat(dataGeneratorJobInfo.state()).isIn(JobState.ACTIVE_STATES);
+    DataflowOperator.Config config = createConfig(dataGeneratorJobInfo, timeout);
+    // check if the job will be BATCH or STREAMING
+    Result dataGeneratorResult;
+    if (dataGeneratorOptions.parameters().containsKey(MESSAGES_LIMIT)) {
+      // BATCH job, wait till data generator job finishes
+      dataGeneratorResult = dataflowOperator.waitUntilDone(config);
+      assertThat(dataGeneratorResult).isEqualTo(Result.JOB_FINISHED);
+    } else {
+      // STREAMING job, wait till timeout and drain job
+      dataGeneratorResult = dataflowOperator.waitUntilDoneAndFinish(config);
+      assertThat(dataGeneratorResult).isEqualTo(Result.TIMEOUT);
+    }
+  }
+
+  /** Builder for the {@link DataGenerator}. */
+  public static final class Builder {
+    private final String jobName;
+    private final Map<String, String> parameters;
+
+    private Builder(String jobName) {
+      this.jobName = jobName;
+      this.parameters = new HashMap<>();
+    }
+
+    public String getJobName() {
+      return jobName;
+    }
+
+    public Map<String, String> getParameters() {
+      return parameters;
+    }
+
+    public DataGenerator.Builder setSchemaLocation(String value) {
+      parameters.put("schemaLocation", value);
+      return this;
+    }
+
+    public DataGenerator.Builder setMessagesLimit(String value) {
+      parameters.put(MESSAGES_LIMIT, value);
+      return this;
+    }
+
+    public DataGenerator.Builder setQPS(String value) {
+      parameters.put("qps", value);
+      return this;
+    }
+
+    public Builder setWorkerMachineType(String value) {
+      parameters.put("workerMachineType", value);
+      return this;
+    }
+
+    public Builder setNumWorkers(String value) {
+      parameters.put("numWorkers", value);
+      return this;
+    }
+
+    public DataGenerator.Builder setMaxNumWorkers(String value) {
+      parameters.put("maxNumWorkers", value);
+      return this;
+    }
+
+    public DataGenerator.Builder setAutoscalingAlgorithm(String value) {
```

Comment:
nit We could use SDK enum AutoscalingAlgorithmType here

Reply:
Done